pkg/scene/widget_align_system.go

```diff
@@ -0,0 +1,200 @@
+
+	switch section {
+	case "left":
+		return &z.left
+	case "center":
+		return &z.center
+	case "right":
+		return &z.right
+	}
+	return nil
+}
+
+// WidgetArea will return a specific area in the align system.
+func (was *WidgetAlignSystem) WidgetArea(zone, section, area string) *WidgetArea {
+	if was == nil {
+		return nil
+	}
+
+	s := was.WidgetSection(zone, section)
+
+	switch area {
+	case "top":
+		return &s.top
+	case "middle":
+		return &s.middle
+	case "bottom":
+		return &s.bottom
+	}
+	return nil
+}
+
+// WidgetIds will return a slice of widget ids from a specific area.
+func (was *WidgetAlignSystem) WidgetIds(z, s, a string) []*id.WidgetID {
+	area := was.WidgetArea(z, s, a)
+	return area.widgetIds
+}
+
+// Alignment will return the alignment of a specific area.
+func (was *WidgetAlignSystem) Alignment(z, s, a string) *string {
+	area := was.WidgetArea(z, s, a)
+	return &area.align
+}
+
+// Add a widget to the align system.
+func (was *WidgetAlignSystem) Add(wid *id.WidgetID, l *WidgetLocation) {
+	if was == nil {
+		return
+	}
+	a := was.WidgetArea(l.Zone, l.Section, l.Area)
+	nIds := a.widgetIds
+	if !a.Has(*wid) {
+		nIds = append(a.widgetIds, wid)
+	}
+	a.widgetIds = nIds
+	if a.align == "" {
+		a.align = "start"
+	}
+}
+
+// Remove a widget from the align system.
+func (was *WidgetAlignSystem) Remove(wid *id.WidgetID, l *WidgetLocation) {
+	if was == nil {
+		return
+	}
+
+	a := was.WidgetArea(l.Zone, l.Section, l.Area)
+
+	var nwid []*id.WidgetID
+	for _, w := range a.widgetIds {
+		if w.ID() != wid.ID() {
+			nwid = append(nwid, w)
+		}
+	}
+	a.widgetIds = nwid
+}
+
+// Update a widget in the align system.
+func (was *WidgetAlignSystem) Update(wid *id.WidgetID, l, newL *WidgetLocation, index, newIndex *int, align *string) {
+	if was == nil && wid == nil && l == nil {
+		return
+	}
+
+	a := was.WidgetArea(l.Zone, l.Section, l.Area)
+
+	if align != nil {
+		a.align = *align
+	} else {
+		a.align = "start"
+	}
+
+	if index != nil && newIndex != nil {
+		moveInt(a.widgetIds, *index, *newIndex)
+	}
+	if newL != nil {
+		was.Remove(wid, l)
+		was.Add(wid, newL)
+	}
+}
+
+// moveInt moves a widget's index.
+func moveInt(array []*id.WidgetID, srcIndex int, dstIndex int) []*id.WidgetID {
+	value := array[srcIndex]
+	return insertInt(removeInt(array, srcIndex), value, dstIndex)
+}
+
+// insertInt is used in moveInt to add the widgetID to a new position(index).
+func insertInt(array []*id.WidgetID, value *id.WidgetID, index int) []*id.WidgetID {
+	return append(array[:index], append([]*id.WidgetID{value}, array[index:]...)...)
+}
+
+// removeInt is used in moveInt to remove the widgetID from original position(index).
+func removeInt(array []*id.WidgetID, index int) []*id.WidgetID {
+	return append(array[:index], array[index+1:]...)
+}
+
+// WidgetAreaFrom will add a slice of widget ids to a specific area of an align system.
+func (was *WidgetAlignSystem) WidgetAreaFrom(wids []*id.WidgetID, align, z, s, a string) {
```

Comment:
```suggestion
func (was *WidgetAlignSystem) AddArea(wids []*id.WidgetID, align, z, s, a string) {
```

A function named From is expected to initialize and return some value, just like the New function.

Reply:
Ok, I misunderstood the naming system! But this function actually adds multiple widgets to an Area, not adding an Area, so how about renaming it AddMultiple?

Reply:
But I am using it in mongodoc to take mongo's widget align system's widgetIds and add them to the struct widget align system so I am not sure that AddMultiple is appropriate either.

Reply:
ok just `AddAll` is good

```suggestion
func (was *WidgetAlignSystem) AddAll(wids []*id.WidgetID, z, s, a string) {
```

Note: align is not used

Reply:
Ah, missed align. This method is currently used in the conversion from mongodoc types to domain models so it is good to convert align as well :P 
Thank you.